Use ArgumentException.ThrowIfNullOrWhitespace

wcontayon opened the pull request:
# Use ArgumentException.ThrowIfNullOrWhitespace

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

## Description
Use **`ArgumentException.ThrowIfNullOrWhitespace`** method from https://github.com/dotnet/runtime/pull/86007
Fixes #48185  (partially, still check if we can replace some code with `ArgumentException.ThrowIfNullOrWhitespace`)

wcontayon committed:
use throwifnullorwhitespace

wcontayon committed:
replace by argumentexception.throwifnullorwhitespace

wcontayon committed:
replace test by argument.throwifnullorempty

mgravell commented on src/Hosting/Hosting/src/Internal/HostingEnvironmentExtensions.cs:
```diff
@@ -13,10 +13,8 @@ internal static void Initialize(this IHostingEnvironment hostingEnvironment, str
-        {
-            throw new ArgumentException("A valid non-empty content root must be provided.", nameof(contentRootPath));
-        }
+        ArgumentException.ThrowIfNullOrEmpty(contentRootPath);
```
Definitely less explanatory

mgravell commented on src/Components/Server/src/ProtectedBrowserStorage/ProtectedBrowserStorage.cs:
```diff
@@ -71,15 +68,8 @@ public ValueTask SetAsync(string key, object value)
-        {
-            throw new ArgumentException("Cannot be null or empty", nameof(key));
-        }
+        ArgumentException.ThrowIfNullOrEmpty(purpose);
```
Observable change - not saying good or bad, but: definitely a change; I'm fine with it personally

danmoseley replied:
I'm missing what observable here?

mgravell replied:
The specific error message will change. Very much a pedantic "we should at least be aware that we're changing something, even if it is something that nobody should reasonably rely on", and in this case very much in the "trivial, no custom messages, just go for it" category.

mgravell replied:
Also note param name that gets blamed: "key" vs "purpose"

wcontayon replied:
Effectively, we do not have an overload that accepts a custom message.
I just open a [Api Proposal on dotnet/runtime](https://github.com/dotnet/runtime/issues/86833) to have the option to set the custom message.
we can just update only those part with generic message

danmoseley replied:
Ah, right. By precedent, I think we've had no problem changing exception messages, including in cases like this, so long as that doesn't lose actual useful information.

mgravell commented on src/Hosting/Hosting/src/Internal/HostingEnvironmentExtensions.cs:
```diff
@@ -67,10 +65,8 @@ internal static void Initialize(this IHostingEnvironment hostingEnvironment, str
-        {
-            throw new ArgumentException("A valid non-empty content root must be provided.", nameof(contentRootPath));
-        }
+        ArgumentException.ThrowIfNullOrEmpty(contentRootPath);
```
Ditto less explanatory

mgravell commented:
Not going to comment on each and every one of these, but I think there are 3 main scenarios in here:

- trivial, no custom messages, just go for it
- vague custom message; observable change, but... not a problem (and probably desirable for consistency)
- humanized, context-specific message, that is erased here

The first two categories: IMO absolutely fine, let's encourage that. It is this final category that worries me a little - some of the existing message IMO cross an invisible and subjective line where the current message is more helpful than the new.

Not at a PC to check, but: I assume there isn't an overload that accepts a custom message?

comlanwcontayon committed:
keep custom message on exception

danmoseley commented on src/Hosting/Hosting/src/Internal/HostingEnvironmentExtensions.cs:
```diff
@@ -13,7 +13,11 @@ internal static void Initialize(this IHostingEnvironment hostingEnvironment, str
         ArgumentNullException.ThrowIfNull(options);
-        ArgumentException.ThrowIfNullOrEmpty(contentRootPath);
+
+       if (string.IsNullOrEmpty(contentRootPath))
```
```suggestion
        if (string.IsNullOrEmpty(contentRootPath))
```

danmoseley commented on src/Http/Routing/src/DefaultLinkGenerator.cs:
```diff
@@ -152,8 +152,15 @@ internal sealed partial class DefaultLinkGenerator : LinkGenerator, IDisposable
     {
-        ArgumentException.ThrowIfNullOrEmpty(scheme);
-        ArgumentException.ThrowIfNullOrEmpty(host.Value, nameof(host));
+         if (string.IsNullOrEmpty(scheme))
```
```suggestion
        if (string.IsNullOrEmpty(scheme))
```

wcontayon committed:
Update src/Http/Routing/src/DefaultLinkGenerator.cs

Co-authored-by: Dan Moseley <danmose@microsoft.com>

danmoseley commented on src/Security/CookiePolicy/src/CookiePolicyOptions.cs:
```diff
@@ -59,10 +59,7 @@ public string ConsentCookieValue
-            {
-                throw new ArgumentException("Value cannot be null or empty string.", nameof(value));
-            }
+            ArgumentException.ThrowIfNullOrEmpty(value, nameof(value));
```
the 2nd parameter is required here?

danmoseley replied:
```suggestion
            ArgumentException.ThrowIfNullOrEmpty(value);
```

danmoseley commented on src/Mvc/Mvc.RazorPages/src/ApplicationModels/PageConventionCollection.cs:
```diff
@@ -275,10 +260,7 @@ internal static void EnsureValidPageName(string pageName, string argumentName =
-        {
-            throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(folderPath));
-        }
+        ArgumentException.ThrowIfNullOrEmpty(folderPath, nameof(folderPath));
```
```suggestion
        ArgumentException.ThrowIfNullOrEmpty(folderPath);
```

danmoseley commented on src/Middleware/Spa/SpaServices.Extensions/src/StaticFiles/DefaultSpaStaticFileProvider.cs:
```diff
@@ -20,12 +20,7 @@ internal sealed class DefaultSpaStaticFileProvider : ISpaStaticFileProvider
-            throw new ArgumentException($"The {nameof(options.RootPath)} property " +
-                $"of {nameof(options)} cannot be null or empty.");
-        }
+        ArgumentException.ThrowIfNullOrEmpty(options.RootPath, nameof(options.RootPath));
```
does this require the 2nd parameter to get it right?

danmoseley replied:
```suggestion
        ArgumentException.ThrowIfNullOrEmpty(options.RootPath);
```

danmoseley commented on src/Http/Routing/src/Patterns/RoutePatternFactory.cs:
```diff
@@ -902,8 +880,6 @@ public static RoutePatternParameterPart ParameterPart(string parameterName, obje
             throw new ArgumentException(Resources.TemplateRoute_OptionalCannotHaveDefaultValue);
         }
 
-        ArgumentNullException.ThrowIfNull(parameterPolicies);
```
were these intentionally moved up?

danmoseley commented on src/Http/Headers/src/CookieHeaderValue.cs:
```diff
@@ -164,10 +164,8 @@ public static bool TryParseStrictList(IList<string>? inputs, [NotNullWhen(true)]
-        {
-            throw new ArgumentNullException(nameof(name));
-        }
+        ArgumentNullException.ThrowIfNull(name, parameterName);
```
This is covered by the line below, and the implicit equals operator on the StringSegment struct, I think? Anyway in other cases you don't have this line.
```suggestion
```

danmoseley commented:
Just a few bits of feedback. Have you verified in the debugger that it's working as expected on StringSegment? Normally you can't compare structs to null but I guess it's using the implicit equality operator.

BTW, there's still 6 uses of `Resources.ArgumentCannotBeNullOrEmpty` in the tree. Do you want to remove those, and then remove the string as well?

Tratcher commented on src/Http/Headers/src/CookieHeaderValue.cs:
```diff
@@ -177,10 +175,7 @@ internal static void CheckNameFormat(StringSegment name, string parameterName)
-        {
-            throw new ArgumentNullException(nameof(value));
-        }
+        ArgumentException.ThrowIfNullOrWhiteSpace(value.Value);
```
Behavior change, Why is this more restrictive than before? This might explain most of the auth test failures, an empty value is commonly used when removing a cookie.

wcontayon commented:
/azp run

danmoseley commented:
@wcontayon To rerun validation you have to click close then click open again..

danmoseley commented:
> https://github.com/dotnet/aspnetcore/pull/48414#issuecomment-1598042342

@mkArtakMSFT this comment seems to only work for folks with write access -- should we clarify the text that other folks should close/reopen? (my assumption is that it runs on latest HEAD either way @wtgodbe ?)

danmoseley commented:
meanwhile, @wcontayon did you have a chance to address the feedback? You can click resolve when you have.
If it makes things easier, your option, you could split the PR up.

wtgodbe commented:
> my assumption is that it runs on latest HEAD either way @wtgodbe ?

That's right

danmoseley commented:
@wcontayon do you expect to be able to continue addressing feedback? thanks for doing this. BTW, if it's easier for you, you could choose to split it into smaller PR's.

wcontayon commented:
> @wcontayon do you expect to be able to continue addressing feedback? thanks for doing this. BTW, if it's easier for you, you could choose to split it into smaller PR's.

@danmoseley yes sure !!
I’ll try to push the update by the end of next week 🙂

danmoseley commented:
Cool! Work on your own schedule, just want to make sure we end up merging your work.

danmoseley commented:
@wcontayon do you plan to continue with this PR? I'll close it for now, but feel free to reopen when you're ready.
